File: Connectors/InteractiveBrokers/Java/Client/ScannerSubscription.java
/* Copyright (C) 2013 Interactive Brokers LLC. All rights reserved.  This code is subject to the terms
 * and conditions of the IB API Non-Commercial License or the IB API Commercial License, as applicable. */

package com.ib.client;

public class ScannerSubscription {
    public final static int NO_ROW_NUMBER_SPECIFIED = -1;

    private int m_numberOfRows = NO_ROW_NUMBER_SPECIFIED;
    private String m_instrument;
    private String m_locationCode;
    private String m_scanCode;
    private double m_abovePrice = Double.MAX_VALUE;
    private double m_belowPrice = Double.MAX_VALUE;
    private int m_aboveVolume = Integer.MAX_VALUE;
    private int m_averageOptionVolumeAbove = Integer.MAX_VALUE;
    private double m_marketCapAbove = Double.MAX_VALUE;
    private double m_marketCapBelow = Double.MAX_VALUE;
    private String m_moodyRatingAbove;
    private String m_moodyRatingBelow;
    private String m_spRatingAbove;
    private String m_spRatingBelow;
    private String m_maturityDateAbove;
    private String m_maturityDateBelow;
    private double m_couponRateAbove = Double.MAX_VALUE;
    private double m_couponRateBelow = Double.MAX_VALUE;
    private String m_excludeConvertible;
    private String m_scannerSettingPairs;
    private String m_stockTypeFilter;

    // Get
    public int numberOfRows()                   { return m_numberOfRows; }
    public String instrument()                  { return m_instrument; }
    public String locationCode()                { return m_locationCode; }
    public String scanCode()                    { return m_scanCode; }
    public double abovePrice()                  { return m_abovePrice; }
    public double belowPrice()                  { return m_belowPrice; }
    public int aboveVolume()                    { return m_aboveVolume; }
    public int averageOptionVolumeAbove()       { return m_averageOptionVolumeAbove; }
    public double marketCapAbove()              { return m_marketCapAbove; }
    public double marketCapBelow()              { return m_marketCapBelow; }
    public String moodyRatingAbove()            { return m_moodyRatingAbove; }
    public String moodyRatingBelow()            { return m_moodyRatingBelow; }
    public String spRatingAbove()               { return m_spRatingAbove; }
    public String spRatingBelow()               { return m_spRatingBelow; }
    public String maturityDateAbove()           { return m_maturityDateAbove; }
    public String maturityDateBelow()           { return m_maturityDateBelow; }
    public double couponRateAbove()             { return m_couponRateAbove; }
    public double couponRateBelow()             { return m_couponRateBelow; }
    public String excludeConvertible()          { return m_excludeConvertible; }
    public String scannerSettingPairs()         { return m_scannerSettingPairs; }
    public String stockTypeFilter()             { return m_stockTypeFilter; }

    // Set
    public void numberOfRows(int num)          { m_numberOfRows = num; }
    public void instrument(String txt)         { m_instrument = txt; }
    public void locationCode(String txt)       { m_locationCode = txt; }
    public void scanCode(String txt)           { m_scanCode = txt; }
    public void abovePrice(double price)       { m_abovePrice = price; }
    public void belowPrice(double price)       { m_belowPrice = price; }
    public void aboveVolume(int volume)        { m_aboveVolume = volume; }
    public void averageOptionVolumeAbove(int volume) { m_averageOptionVolumeAbove = volume; }
    public void marketCapAbove(double cap)     { m_marketCapAbove = cap; }
    public void marketCapBelow(double cap)     { m_marketCapBelow = cap; }
    public void moodyRatingAbove(String r)     { m_moodyRatingAbove = r; }
    public void moodyRatingBelow(String r)     { m_moodyRatingBelow = r; }
    public void spRatingAbove(String r)        { m_spRatingAbove = r; }
    public void spRatingBelow(String r)        { m_spRatingBelow = r; }
    public void maturityDateAbove(String d)    { m_maturityDateAbove = d; }
    public void maturityDateBelow(String d)    { m_maturityDateBelow = d; }
    public void couponRateAbove(double r)      { m_couponRateAbove = r; }
    public void couponRateBelow(double r)      { m_couponRateBelow = r; }
    public void excludeConvertible(String c)   { m_excludeConvertible = c; }
    public void scannerSettingPairs(String val) { m_scannerSettingPairs = val; }
    public void stockTypeFilter(String val)    { m_stockTypeFilter = val; }
}

File: Connectors/InteractiveBrokers/CSharp/ScannerSubscription.cs
/* Copyright (C) 2013 Interactive Brokers LLC. All rights reserved.  This code is subject to the terms
 * and conditions of the IB API Non-Commercial License or the IB API Commercial License, as applicable. */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;

namespace IBApi
{
    /**
     * @class ScannerSubscription
     * @brief Defines a market scanner request
     */
    public class ScannerSubscription
    {
        /**
         * @var int numberOfRows
         * @brief The number of rows to be returned for the query
         */
        private int numberOfRows = -1;

        /**
         * @var string instrument
         * @brief The instrument's type for the scan. I.e. STK, FUT.HK, etc.
         */
        private string instrument;

        /**
         * @var string locationCode
         * @brief The request's location (STK.US, STK.US.MAJOR, etc). 
         */
        private string locationCode;

        /**
         * @var string scanCode
         * @brief Same as TWS Market Scanner's "parameters" field, for example: TOP_PERC_GAIN
         */
        private string scanCode;

        /**
         * @var double abovePrice
         * @brief Filters out Contracts which price is below this value
         */
        private double abovePrice = Double.MaxValue;

        /**
         * @var double belowPrice
         * @brief Filters out contracts which price is above this value.
         */
        private double belowPrice = Double.MaxValue;

        /**
         * @var int aboveVolume
         * @brief Filters out Contracts which volume is above this value.
         */
        private int aboveVolume = Int32.MaxValue;

        /**
         * @var int averageOptionVolumeAbove
         * @brief Filters out Contracts which option volume is above this value.
         */
        private int averageOptionVolumeAbove = Int32.MaxValue;

        /**
         * @var double marketCapAbove
         * @brief Filters out Contracts which market cap is above this value.
         */
        private double marketCapAbove = Double.MaxValue;

        /**
         * @var double marketCapBelow
         * @brief Filters out Contracts which market cap is below this value.
         */
        private double marketCapBelow = Double.MaxValue;

        /**
         * @var string moodyRatingAbove
         * @brief Filters out Contracts which Moody's rating is below this value.
         */
        private string moodyRatingAbove;

        /**
         * @var string moodyRatingBelow
         * @brief Filters out Contracts which Moody's rating is above this value.
         */
        private string moodyRatingBelow;

        /**
         * @var string spRatingAbove
         * @brief Filters out Contracts with a S&P rating below this value.
         */
        private string spRatingAbove;

        /**
         * @var string spRatingBelow
         * @brief Filters out Contracts with a S&P rating above this value.
         */
        private string spRatingBelow;

        /**
         * @var string maturityDateAbove
         * @brief Filter out Contracts with a maturity date earlier than this value.
         */
        private string maturityDateAbove;

        /**
         * @var string maturityDateBelow
         * @brief Filter out Contracts with a maturity date older than this value.
         */
        private string maturityDateBelow;

        /**
         * @var double couponRateAbove
         * @brief Filter out Contracts with a coupon rate lower than this value.
         */
        private double couponRateAbove = Double.MaxValue;

        /**
         * @var double couponRateBelow
         * @brief Filter out Contracts with a coupon rate higher than this value.
         */
        private double couponRateBelow = Double.MaxValue;

        /**
         * @var string excludeConvertible
         * @brief Filters out Convertible bonds
         */
        private string excludeConvertible;

        /**
         * @var string scannerSettingPairs
         * @brief For example, a pairing "Annual, true" used on the "top Option Implied Vol % Gainers" scan would return annualized volatilities.
         */
        private string scannerSettingPairs;

        /**
         * @var string stockTypeFilter
         * @brief -
         *      CORP = Corporation
         *      ADR = American Depositary Receipt
         *      ETF = Exchange Traded Fund
         *      REIT = Real Estate Investment Trust
         *      CEF = Closed End Fund
         */
        private string stockTypeFilter;

        public int NumberOfRows
        {
            get { return numberOfRows; }
            set { numberOfRows = value; }
        }
       
        public string Instrument
        {
            get { return instrument; }
            set { instrument = value; }
        }

        public string LocationCode
        {
            get { return locationCode; }
            set { locationCode = value; }
        }

        public string ScanCode
        {
            get { return scanCode; }
            set { scanCode = value; }
        }

        public double AbovePrice
        {
            get { return abovePrice; }
            set { abovePrice = value; }
        }

        public double BelowPrice
        {
            get { return belowPrice; }
            set { belowPrice = value; }
        }

        public int AboveVolume
        {
            get { return aboveVolume; }
            set { aboveVolume = value; }
        }

        public int AverageOptionVolumeAbove
        {
            get { return averageOptionVolumeAbove; }
            set { averageOptionVolumeAbove = value; }
        }

        public double MarketCapAbove
        {
            get { return marketCapAbove; }
            set { marketCapAbove = value; }
        }

        public double MarketCapBelow
        {
            get { return marketCapBelow; }
            set { marketCapBelow = value; }
        }

        public string MoodyRatingAbove
        {
            get { return moodyRatingAbove; }
            set { moodyRatingAbove = value; }
        }

        public string MoodyRatingBelow
        {
            get { return moodyRatingBelow; }
            set { moodyRatingBelow = value; }
        }

        public string SpRatingAbove
        {
            get { return spRatingAbove; }
            set { spRatingAbove = value; }
        }

        public string SpRatingBelow
        {
            get { return spRatingBelow; }
            set { spRatingBelow = value; }
        }

        public string MaturityDateAbove
        {
            get { return maturityDateAbove; }
            set { maturityDateAbove = value; }
        }

        public string MaturityDateBelow
        {
            get { return maturityDateBelow; }
            set { maturityDateBelow = value; }
        }

        public double CouponRateAbove
        {
            get { return couponRateAbove; }
            set { couponRateAbove = value; }
        }

        public double CouponRateBelow
        {
            get { return couponRateBelow; }
            set { couponRateBelow = value; }
        }

        public string ExcludeConvertible
        {
            get { return excludeConvertible; }
            set { excludeConvertible = value; }
        }

        public string ScannerSettingPairs
        {
            get { return scannerSettingPairs; }
            set { scannerSettingPairs = value; }
        }

        public string StockTypeFilter
        {
            get { return stockTypeFilter; }
            set { stockTypeFilter = value; }
        }
    }
}
